pkg/frontend/querymiddleware/error_caching.go

```diff
@@ -0,0 +1,182 @@
+	spanLog := spanlogger.FromContext(ctx, e.logger)
+	tenantIDs, err := tenant.TenantIDs(ctx)
+	if err != nil {
+		return e.next.Do(ctx, request)
+	}
+
+	// Check if caching has disabled via an option on the request
+	if !e.shouldCacheReq(request) {
+		e.cacheStoreSkipped.WithLabelValues(reasonDisabledByOption).Inc()
+		return e.next.Do(ctx, request)
+	}
+
+	e.cacheAttempted.Inc()
+	key := e.cacheKey(tenant.JoinTenantIDs(tenantIDs), request)
+	hashedKey := cacheHashKey(key)
+
+	if cachedErr := e.loadErrorFromCache(ctx, key, hashedKey); cachedErr != nil {
+		e.cacheHits.Inc()
+		spanLog.DebugLog(
+			"msg", "returned cached API error",
+			"error_type", cachedErr.Type,
+			"key", key,
+			"hashed_key", hashedKey,
+		)
+
+		return nil, cachedErr
+	}
+
+	res, err := e.next.Do(ctx, request)
+	if err != nil {
+		var apiErr *apierror.APIError
+		if !errors.As(err, &apiErr) {
+			e.cacheStoreSkipped.WithLabelValues(reasonNotAPIError).Inc()
+			return res, err
+		}
+
+		if cacheable, reason := e.isCacheable(apiErr, request, tenantIDs); !cacheable {
+			e.cacheStoreSkipped.WithLabelValues(reason).Inc()
+			spanLog.DebugLog(
+				"msg", "error result from request is not cacheable",
+				"error_type", apiErr.Type,
+				"reason", reason,
+			)
+			return res, err
+		}
+
+		e.storeErrorToCache(key, hashedKey, apiErr)
+	}
+
+	return res, err
+}
+
+func (e *errorCachingHandler) loadErrorFromCache(ctx context.Context, key, hashedKey string) *apierror.APIError {
+	res := e.cache.GetMulti(ctx, []string{hashedKey})
+	if cached, ok := res[hashedKey]; ok {
+		var cachedError CachedError
+		if err := proto.Unmarshal(cached, &cachedError); err != nil {
+			level.Warn(e.logger).Log("msg", "unable to unmarshall cached error", "err", err)
+			return nil
+		}
+
+		if cachedError.GetKey() != key {
+			level.Debug(e.logger).Log(
+				"msg", "cached error key does not match",
+				"expected_key", key,
+				"actual_key", cachedError.GetKey(),
+				"hashed_key", hashedKey,
+			)
+			return nil
+		}
+
+		return apierror.New(apierror.Type(cachedError.Type), cachedError.Message)
+	}
+
+	return nil
+}
+
+func (e *errorCachingHandler) storeErrorToCache(key, hashedKey string, apiErr *apierror.APIError) {
+	bytes, err := proto.Marshal(&CachedError{
+		Key:     key,
+		Type:    string(apiErr.Type),
+		Message: apiErr.Message,
+	})
+
+	if err != nil {
+		level.Warn(e.logger).Log("msg", "unable to marshal cached error", "err", err)
+		return
+	}
+
+	e.cache.SetAsync(hashedKey, bytes, e.ttl)
+}
+
+func (e *errorCachingHandler) cacheKey(tenantID string, r MetricsQueryRequest) string {
+	return fmt.Sprintf("EC:%s:%s:%d:%d:%d", tenantID, r.GetQuery(), r.GetStart(), r.GetEnd(), r.GetStep())
+}
+
+func (e *errorCachingHandler) isCacheable(apiErr *apierror.APIError, req MetricsQueryRequest, tenantIDs []string) (bool, string) {
+	if apiErr.Type != apierror.TypeBadData && apiErr.Type != apierror.TypeExec {
```

Comment:
I saw we also have HTTP 413 responses. I couldn't find what they're used for though. Should we cache them too? https://github.com/grafana/mimir/blob/e78426575885594438de073d8221f544b92add62/pkg/frontend/querymiddleware/codec.go#L724-L725

Reply:
I think that's reasonable. I'll take a look and try to find some examples.